flytekit/image_spec/default_builder.py

```diff
@@ -0,0 +1,238 @@
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+import warnings
+from pathlib import Path
+from string import Template
+from typing import ClassVar, Optional
+
+import click
+
+from flytekit.image_spec.image_spec import _F_IMG_ID, ImageSpec, ImageSpecBuilder
+
+# The default image builder base image is located at Dockerfile.default-image-builder.
+# For local testing, build the base image by running:
+# 1. Set environment variable `DEFAULT_BUILDER_BASE_IMAGE=localhost:30000/default-image-builder-base`
+# 2. make setup-multiarch-builder
+# 3. make build-default-image-builder-image
+DEFAULT_BUILDER_BASE_IMAGE_ENV = "DEFAULT_BUILDER_BASE_IMAGE"
+DEFAULT_BUILDER_BASE_IMAGE = "thomasjpfan/default-image-builder-base:0.0.3"
+
+BASE_IMAGE_BUILDER = os.getenv(DEFAULT_BUILDER_BASE_IMAGE_ENV, DEFAULT_BUILDER_BASE_IMAGE)
+
+
+PYTHON_INSTALL_COMMAND = """\
+RUN --mount=type=cache,target=/root/.cache/uv,id=uv \
+    --mount=type=bind,target=requirements.txt,src=requirements.txt \
+    /opt/conda/bin/uv \
+    pip install --python /opt/conda/envs/dev/bin/python $PIP_INDEX \
+    --requirement requirements.txt
+"""
+
+APT_INSTALL_COMMAND_TEMPLATE = Template(
+    """\
+RUN --mount=type=cache,target=/var/cache/apt,id=apt \
+    apt-get update && apt-get install -y --no-install-recommends \
+    $APT_PACKAGES
+"""
+)
+
+DOCKER_FILE_TEMPLATE = Template(
+    """\
+#syntax=docker/dockerfile:1.5
+FROM $BASE_IMAGE_BUILDER as build
+
+RUN --mount=type=cache,target=/opt/conda/pkgs,id=conda \
+    mamba create \
+        -c conda-forge $CONDA_CHANNELS \
+        -n dev -y python=$PYTHON_VERSION $CONDA_PACKAGES
+
+WORKDIR /root
+
+$COPY_COMMAND_BUILDER
+
+$PYTHON_INSTALL_COMMAND
+
+RUN /opt/conda/bin/conda-pack -n dev -o /tmp/env.tar && \
+    mkdir /venv && cd /venv && tar xf /tmp/env.tar && \
+    rm /tmp/env.tar
+
+RUN /venv/bin/conda-unpack
+
+FROM $BASE_IMAGE AS runtime
+
+RUN --mount=type=cache,target=/var/cache/apt,id=apt \
+    apt-get update && apt-get install -y --no-install-recommends \
+    ca-certificates
+RUN update-ca-certificates
+
+$APT_INSTALL_COMMAND
+
+COPY --from=build /venv /venv
+ENV PATH="/venv/bin:$$PATH"
+
+WORKDIR /root
+ENV FLYTE_SDK_RICH_TRACEBACKS=0 SSL_CERT_DIR=/etc/ssl/certs $ENV
+
+RUN useradd --create-home --shell /bin/bash -u 1000 flytekit \
+    && chown -R flytekit /root \
+    && chown -R flytekit /home
+
+$COPY_COMMAND_RUNTIME
+
+$RUN_COMMANDS
+
+RUN echo "source /venv/bin/activate" >> /home/flytekit/.bashrc
+SHELL ["/bin/bash", "-c"]
+
+USER flytekit
+"""
+)
+
+
+def create_docker_context(image_spec: ImageSpec, tmp_dir: Path):
+    """Populate tmp_dir with Dockerfile as specified by the `image_spec`."""
+    base_image = image_spec.base_image or "debian:bookworm-slim"
+
+    if image_spec.cuda:
+        # Base image requires an NVIDIA driver. cuda and cudnn will be installed with conda
```

Comment:
The semantics here are not great. If we install `cuda` with `conda` and then install pytorch through `pypi`, then `cuda` will be installed twice.

- PyTorch through pypi will pull in cuda from PyPI
- Using conda to install cuda adds a second one

We can add more logic to do: If `cuda` is set:

- Install pypi packages
- Check if cuda is installed
    - If pypi's cuda is not installed, get it from it from conda (or pypi)
    - If pypi's cuda is installed, do nothing.

Reply:
Agree, should we deprecate `cuda`, `cudnn`?
Thus, if people want to use cuda, they do

```python
ImageSpec(
  base_image="nvidia/cuda:12.2.2-cudnn8-runtime-ubi8", 
  packages=["flytekit"]
)
```
or 
```python
ImageSpec(
  packages=["cudnn"]
)
```
wdyt

Reply:
With my recent changes, if people use `cuda` with python libraries, then do not need to specify anything extra in `ImageSpec`. If they install `pytorch` from `pypi` or `conda`, then it already comes with cuda.

Their Kubernetes setup needs to be configured to inject the GPU driver into the container.